eslint/eslint-patch/src/eslint-bulk-suppressions/bulk-suppressions-patch.ts

```diff
@@ -1,279 +1,367 @@
+  suppressionsConfig.jsonObject.suppressions.sort(compareSuppressions);
+  fs.writeFileSync(suppressionsPath, JSON.stringify(suppressionsConfig.jsonObject, undefined, 2));
 }
 
-const usedSuppressions: Set<string> = new Set<string>();
-
-function serializeSuppression(fileAbsolutePath: string, suppression: ISuppression): string {
-  return `${fileAbsolutePath}|${suppression.file}|${suppression.scopeId}|${suppression.rule}`;
+function serializeSuppression({ file, scopeId, rule }: ISuppression): string {
+  return `${file}|${scopeId}|${rule}`;
 }
 
 // One-line insert into the ruleContext report method to prematurely exit if the ESLint problem has been suppressed
 export function shouldBulkSuppress(params: {
   filename: string;
   currentNode: TSESTree.Node;
   ruleId: string;
+  problem: IProblem;
 }): boolean {
   // Use this ENV variable to turn off eslint-bulk-suppressions functionality, default behavior is on
-  if (process.env.ESLINT_BULK_ENABLE === 'false') return false;
+  if (process.env[ESLINT_BULK_ENABLE_ENV_VAR_NAME] === 'false') {
+    return false;
+  }
 
-  const { filename: fileAbsolutePath, currentNode, ruleId: rule } = params;
-  const eslintrcDirectory: string = findEslintrcDirectory(fileAbsolutePath);
-  const fileRelativePath: string = path.relative(eslintrcDirectory, fileAbsolutePath);
+  const { filename: fileAbsolutePath, currentNode, ruleId: rule, problem } = params;
+  const normalizedFileAbsolutePath: string = fileAbsolutePath.replace(/\\/g, '/');
+  const eslintrcDirectory: string =
+    findEslintrcFolderPathForNormalizedFileAbsolutePath(normalizedFileAbsolutePath);
+  const fileRelativePath: string = normalizedFileAbsolutePath.substring(eslintrcDirectory.length + 1);
   const scopeId: string = calculateScopeId(currentNode);
   const suppression: ISuppression = { file: fileRelativePath, scopeId, rule };
 
-  if (shouldWriteSuppression(fileAbsolutePath, suppression)) {
-    writeSuppressionToFile(fileAbsolutePath, suppression);
-  }
-
-  const shouldBulkSuppress: boolean = isSuppressed(fileAbsolutePath, suppression);
+  const config: IBulkSuppressionsConfig = getSuppressionsConfigForEslintrcFolderPath(eslintrcDirectory);
+  const serializedSuppression: string = serializeSuppression(suppression);
+  const currentNodeIsSuppressed: boolean = config.serializedSuppressions.has(serializedSuppression);
 
-  if (shouldBulkSuppress) {
-    usedSuppressions.add(serializeSuppression(fileAbsolutePath, suppression));
+  if (currentNodeIsSuppressed || SUPPRESS_ALL_RULES || RULES_TO_SUPPRESS?.has(suppression.rule)) {
+    problem[SUPPRESSION_SYMBOL] = {
+      suppression,
+      serializedSuppression,
+      config
+    };
   }
 
-  return shouldBulkSuppress;
+  return process.env[ESLINT_BULK_PRUNE_ENV_VAR_NAME] !== '1' && currentNodeIsSuppressed;
 }
 
-export function onFinish(params: { filename: string }): void {
-  if (process.env.ESLINT_BULK_PRUNE === 'true') {
-    bulkSuppressionsPrune(params);
+export function prune(): void {
```

Comment:
To clarify, does this fundamentally change the behavior of the prune operation? 

Originally it was intended to only remove suppressions that no longer did anything (e.g. the source code of the ESLint error was removed).

It looks like this implementation just starts the bulk suppressions file from scratch? I think this is okay too, but if my understanding is correct, it deviates from the original intention, and we should update the docs accordingly.

Reply:
The behavior should be the same, with the exception that issues that were suppressed via comments are now removed from the bulk suppression file.

It does start the suppressions file over, but it only records entries that existed in the original file and were encountered by eslint again.